src/plugins/home/public/application/application.tsx

```diff
@@ -44,7 +44,7 @@ export const renderApp = async (
       );
 
     render(
-      <KibanaRenderContextProvider i18n={coreStart.i18n} theme={coreStart.theme}>
+      <KibanaRenderContextProvider {...coreStart}>
```

Comment:
Not sure if there are other places in this PR to improve this, but I think this is the best way we can correctly pass these dependencies. the more often we do like this the better (less places to update in case we need something else from core) 👍

Reply:
There is a place where I think this wouldn't work very well, but I'm not sure and would like to get your thoughts. Say we have this in the code:

```typescript
const showErrorToast = useCallback((errorTitle: string, error: Error) => {
  toasts.addDanger({
    title: errorTitle,
    text: toMountPoint(<Markdown>{error.message}</Markdown>, core);
  });
}, [toasts, core]);
```

Explanation: the `toMountPoint` function requires services from `coreStart` as a second parameter, because it creates a new rendering context. In `useCallback`, etc, the second parameter is an array of dependencies needed in the body of the callback. If a dependency changes, then the component re-renders. It could negatively impact performance if one of the dependencies is a large complex object. So the above code should change to:

```typescript
const showErrorToast = useCallback((errorTitle: string, error: Error) => {
  toasts.addDanger({
    title: errorTitle,
    text: toMountPoint(<Markdown>{error.message}</Markdown>, { i18n, theme, analytics, userProfile });
  });
}, [toasts, i18n, theme, analytics, userProfile]);
```

Thoughts? @Dosant @sebelga @eokoneyo @clintandrewhall

Reply:
This is an interesting example, but I think it's an edge case and is exempt from the pattern.

`useCallback`, `useEffect` and others _always_ require forethought, and the second parameter is strongly typed.  Still, we can code defensively within the contexts to handle this case.

We should destructure the incoming parameter in the `KibanaRootContextProvider` and perhaps also in the `KibanaRenderContextProvider`.  This would effectively "discard" unused portions of `core`, before adding them to the context value, but still allow people to pass the entire construct if we were to need something else

[KibanaRootContextProvider](https://github.com/elastic/kibana/blob/main/packages/react/kibana_context/root/root_provider.tsx#L58) before:

```
    return (
      <KibanaEuiProvider {...props}>
        <i18n.Context>{children}</i18n.Context>
      </KibanaEuiProvider>
    );
  }
```

After:

```
    const {
      i18n,
      theme,
      userProfile,
      globalStyles
    } = props;

    return (
      <KibanaEuiProvider {...{ theme, userProfile, globalStyles }}>
        <i18n.Context>{children}</i18n.Context>
      </KibanaEuiProvider>
    );
  }
```

Reply:
> This is an interesting example, but I think it's an edge case and is exempt from the pattern.

Thanks @clintandrewhall. I was interested in thoughts/feedback about whether my understanding my understanding of `useCallback` performance is valid, and it sounds like you are saying that it is.

Reply:
I am, but I'm _also_ saying we should destructure the incoming parameters on the contexts.

Typescript doesn't prevent extraneous properties from being set to the context, so we should be sure to extract only those properties necessary.

Reply:
Thanks @clintandrewhall. That is a solid observation and one that I can easily follow up with. 

On the original subject of the performance concern with `useCallback` and using large objects, I want to include this message from @Dosant taken from a separate conversation:
> broadly about re-rendering concern, hopefully as we pick up with react upgrades and in a year or two when this thing is stable, there will be a compiler that would optimize rendering so destructuring and new objects will no longer a performance concern https://react.dev/learn/react-compiler